Closes #4120 Add color options for card borders

mrbeiley opened the pull request:
This PR adds a new form element to the Card Deck paragraph type, 'card_deck_border_color'. Setting this field will add a colored border to each of the Cards in the deck. By default, the Cards maintain the normal, Dark Silver border. This PR also adds the ability to set border colors for individual Cards. If both individual and deck border colors are set, the individual card borders will take precedence.

## Related issues
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
- #4120

## How to test
<!--- Please describe in detail how reviewers can test your changes -->
<!--- Include details of your testing environment and the tests you ran -->
1. Create a Card deck with multiple Cards
2. Set value for 'Card deck border color'
3. Check that all Cards in any given deck have the same border color.
4. Create another Card deck with multiple Cards
5. Set different values for 'Card border color'
6. Check that each card has the assigned color
7. Add a 'Card deck border color' to the same Deck
8. Verify the individual border colors display, not the uniform color

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [ ] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [x] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] My change requires release notes.

Commit:
troy help - implementing deck border color

Commit:
Set a single border color for all cards.

Commit:
Set default to normal border.

Commit:
Fix PHP syntax errors.

Commit:
Merge branch 'main' into issue/4120

mrbeiley commented:
Need help determining the potential 'need' for setting either one border color for each Card in a deck, or setting Card border colors individually. If both are needed, we can create a checkbox, swapping between either option. Right now, setting a single color for all Cards in a deck is implemented.

Commit:
Add ability to set individual Card border colors.

Commit:
PHP codesniffer fixes.

Commit:
Merge branch 'main' into issue/4120

Commit:
Fix behavior setting name.

Co-authored-by: Joe Parsons <471936+joeparsons@users.noreply.github.com>

mrbeiley commented on modules/custom/az_card/src/Plugin/Field/FieldWidget/AZCardWidget.php:
```diff
@@ -294,6 +294,26 @@ public function formElement(FieldItemListInterface $items, $delta, array $elemen
+      '#title' => $this->t('Card Border Color'),
+      '#type' => 'select',
+      '#options' => [
+        '' => $this->t('Dark Silver (default)'),
```
I'm trying to think of an appropriate placeholder, since we don't want to add an actual class when selecting "Dark Silver (default)". Maybe something like `dark-silver` or `dark-silver-fake`?

Commit:
Replace empty string with default border color.

Commit:
Fix PHP issues.

Commit:
Merge branch 'main' into issue/4120

danahertzberg commented on modules/custom/az_card/src/Plugin/Field/FieldWidget/AZCardWidget.php:
```diff
@@ -294,6 +294,26 @@ public function formElement(FieldItemListInterface $items, $delta, array $elemen
+      '#title' => $this->t('Card Border Color'),
+      '#type' => 'select',
+      '#options' => [
+        'border-dark-silver' => $this->t('Dark Silver (default)'),
```
```suggestion
        'border' => $this->t('Default'),
```

danahertzberg commented on modules/custom/az_card/src/Plugin/Field/FieldWidget/AZCardWidget.php:
```diff
@@ -294,6 +294,26 @@ public function formElement(FieldItemListInterface $items, $delta, array $elemen
+        'border-warm-gray' => $this->t('Warm Gray'),
+        'border-white' => $this->t('White'),
+      ],
+      '#default_value' => (!empty($item->options['border_color'])) ? $item->options['border_color'] : 'border-dark-silver',
```
```suggestion
      '#default_value' => (!empty($item->options['border_color'])) ? $item->options['border_color'] : 'border',
```

mrbeiley commented:
To Do: If `Borderless cards` is selected, then hide all border related fields from Card widget

Commit:
Change default key.

tadean commented on modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZCardsParagraphBehavior.php:
```diff
@@ -149,6 +149,26 @@ public function buildBehaviorForm(ParagraphInterface $paragraph, array &$form, F
       '#description' => $this->t('Change the color of the Card group title.'),
     ];
 
+    // Card deck border color.
```
@mrbeiley I did some investigating and it looks like the behavior plugins have just enough information in their parents array to compute a state selector for the element. Something like this might work:
```suggestion
    // Compute selector for conditional state from array parents.
    $parents = $form['#parents'] ?? [];
    // We're looking to compare to the card style element.
    $parents[] = 'card_style';
    $name = array_shift($parents);
    if (count($parents)) {
      $name .= '[' . implode('][', $parents) . ']';
    }
    $selector = ':input[name="' . $name . '"]';
    // Card deck border color.
```

tadean commented on modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZCardsParagraphBehavior.php:
```diff
@@ -149,6 +149,26 @@ public function buildBehaviorForm(ParagraphInterface $paragraph, array &$form, F
+    $form['card_deck_border_color'] = [
+      '#title' => $this->t('Card group border color'),
+      '#type' => 'select',
+      '#options' => [
```
```suggestion
      // Hidden for borderless cards.
      '#states' => [
        'visible' => [
          $selector => ['value' => 'card card-borderless'],
        ],
      ],
      '#options' => [
```

Commit:
Hide 'Card group border color' when 'Card style' is "Borderless cards".

Apply suggestions from Troy's code review - Compute selector for conditional state from array parents, add an 'invisible' state to the card deck border color field when borderless cards are selected

Co-authored-by: tadean <51835841+tadean@users.noreply.github.com>

zsaenz commented:
The other option (I think Joshua's is more viable though) is to fully allow each individual card to override the settings. The idea being that the group Border/Borderless and group color are global, but can be overridden individually where needed.
The downside here is it doesn't make sense to select Borderless, then be able to override that with a border. If that is true, the opposite should apply and a bordered card should have a "no border" option.

zsaenz reviewed (changes requested):
Waiting on recent comments + 3x updates for review

mrbeiley committed:
Merge branch 'main' into issue/4120

mrbeiley committed:
Merge branch 'main' into issue/4120
